### PR TITLE
cni: add ConfigureFirewall error propagation

### LINKED ISSUE #5809: cni plugin failing silently on ADD
### What is the issue?
The CNI plugin currently fails silently on ADD command. The corresponding NAT iptables rules are not created on the Pod on startup, but the CNI reports a success and the Pod continues to start up eventually causing 503 failures. 

### How can it be reproduced?
The failures are random, but I have observed that the likelihood of the failure is much higher when making a large number of changes to kubernets objects in a CNI-enabled cluster. Specifically, if I run a `kubectl rollout restart deploy` on my namespace (causing > 10 pods to restart simultaneously with 30 pods in total), there are usually a few pods that display this behavior.

### Logs, error output, etc
CNI Configuration: 
```json
# cat 10-calico.conflist
{
  "name": "k8s-pod-network",
  "cniVersion": "0.3.1",
  "plugins": [
    {
      "type": "calico",
      "mtu": 1460,
      "log_level": "warning",
      "datastore_type": "kubernetes",
      "nodename": "gke-test-test-core-1de1b59f-8gbq",
      "ipam": {
        "type": "host-local",
        "subnet": "usePodCidr"
      },
      "policy": {
        "type": "k8s"
      },
      "kubernetes": {
        "kubeconfig": "/etc/cni/net.d/calico-kubeconfig"
      }
    },
    {
      "type": "portmap",
      "capabilities": {
        "portMappings": true
      },
      "snat": true
    },
    {
      "type": "bandwidth",
      "capabilities": {
        "bandwidth": true
      }
    },
    {
      "name": "linkerd-cni",
      "type": "linkerd-cni",
      "log_level": "debug",
      "policy": {
        "type": "k8s",
        "k8s_api_root": "https://10.252.0.1:443",
        "k8s_auth_token": "SECRET"
      },
      "kubernetes": {
        "kubeconfig": "/etc/cni/net.d/ZZZ-linkerd-cni-kubeconfig"
      },
      "linkerd": {
        "incoming-proxy-port": 4143,
        "outgoing-proxy-port": 4140,
        "proxy-uid": 2102,
        "ports-to-redirect": [],
        "inbound-ports-to-ignore": [
          "25",
          "443",
          "587",
          "3306",
          "11211"
        ],
        "outbound-ports-to-ignore": [
          "25",
          "443",
          "587",
          "3306",
          "11211"
        ],
        "simulate": false,
        "use-wait-flag": false
      }
    }
  ]
}
```

Note: `appA` is the problem pod; `appB` is a good reference pod on the same node set up with the same CNI plugin.

CNI debug log output
```text
time="2021-02-23T06:04:11Z" level=debug msg="linkerd-cni: cmdAdd, config parsed" prevResult="Interfaces:[{Name:calief42b505d5e Mac: Sandbox:}], IP:[{Version:4 Interface:<nil> Address:{IP:10.250.4.38 Mask:ffffffff} Gateway:<nil>}], DNS:{Nameservers:[] Domain: Search:[] Options:[]}" version=0.3.1
time="2021-02-23T06:04:11Z" level=debug msg="linkerd-cni: setting up iptables firewall" ContainerID=ef973ded8904c9450248bccbbd48f6539e94bc768da071a9f2caf097626551c8 Namespace=app Pod=appA-6df8cc584c-lw5km
time="2021-02-23T06:04:11Z" level=debug msg="linkerd-cni: plugin is finished"
time="2021-02-23T06:04:12Z" level=debug msg="linkerd-cni: cmdAdd, config parsed" prevResult="Interfaces:[{Name:calib6d3f022dd2 Mac: Sandbox:}], IP:[{Version:4 Interface:<nil> Address:{IP:10.250.4.41 Mask:ffffffff} Gateway:<nil>}], DNS:{Nameservers:[] Domain: Search:[] Options:[]}" version=0.3.1
time="2021-02-23T06:04:12Z" level=debug msg="linkerd-cni: setting up iptables firewall" ContainerID=d22a9cd3f7b0606956579db792542dc2e7f20b4ef3239bbaa3f173ef7c0f4e73 Namespace=app Pod=appB-5ffd469587-g95bg
time="2021-02-23T06:04:12Z" level=debug msg="linkerd-cni: plugin is **finished"**
```

appA Pod iptables output
```text
/usr/src/app # iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination
```

appB Pod iptables output
```text
root@appB-5ffd469587-g95bg:/# iptables-legacy -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
PROXY_INIT_REDIRECT  all  --  anywhere             anywhere             /* proxy-init/install-proxy-init-prerouting/1614060252 */

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
PROXY_INIT_OUTPUT  all  --  anywhere             anywhere             /* proxy-init/install-proxy-init-output/1614060252 */

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         

Chain PROXY_INIT_OUTPUT (1 references)
target     prot opt source               destination         
PROXY_INIT_REDIRECT  all  --  anywhere            !localhost            owner UID match 2102 /* proxy-init/redirect-non-loopback-local-traffic/1614060252 */
RETURN     all  --  anywhere             anywhere             owner UID match 2102 /* proxy-init/ignore-proxy-user-id/1614060252 */
RETURN     all  --  anywhere             anywhere             /* proxy-init/ignore-loopback/1614060252 */
RETURN     tcp  --  anywhere             anywhere             multiport dports 25,443,587,3306,11211 /* proxy-init/ignore-port-25,443,587,3306,11211/1614060252 */
REDIRECT   tcp  --  anywhere             anywhere             /* proxy-init/redirect-all-outgoing-to-proxy-port/1614060252 */ redir ports 4140

Chain PROXY_INIT_REDIRECT (2 references)
target     prot opt source               destination         
RETURN     tcp  --  anywhere             anywhere             multiport dports 25,443,587,3306,11211 /* proxy-init/ignore-port-25,443,587,3306,11211/1614060252 */
REDIRECT   tcp  --  anywhere             anywhere             /* proxy-init/redirect-all-incoming-to-proxy-port/1614060252 */ redir ports 4143
```

#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-cni-plugin
------------------
√ cni plugin ConfigMap exists
√ cni plugin PodSecurityPolicy exists
√ cni plugin ClusterRole exists
√ cni plugin ClusterRoleBinding exists
√ cni plugin Role exists
√ cni plugin RoleBinding exists
√ cni plugin ServiceAccount exists
√ cni plugin DaemonSet exists
√ cni plugin pod is running on all nodes

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ tap API server has valid cert
√ tap API server cert is valid for at least 60 days
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days

linkerd-identity-data-plane
---------------------------
√ data plane proxies certificate match CA

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ tap api service is running

linkerd-version
---------------
√ can determine the latest version
‼ cli is up-to-date
    is running version 2.9.2 but the latest stable version is 2.9.3
    see https://linkerd.io/checks/#l5d-version-cli for hints

linkerd-data-plane
------------------
√ data plane namespace exists
√ data plane proxies are ready
E0223 11:27:34.859764  156886 portforward.go:385] error copying from local connection to remote stream: read tcp4 127.0.0.1:39359->127.0.0.1:50132: read: connection reset by peer
E0223 11:27:35.107017  156886 portforward.go:385] error copying from local connection to remote stream: read tcp4 127.0.0.1:39359->127.0.0.1:50138: read: connection reset by peer
√ data plane proxy metrics are present in Prometheus
√ data plane is up-to-date
E0223 11:27:35.228480  156886 portforward.go:385] error copying from local connection to remote stream: read tcp4 127.0.0.1:39359->127.0.0.1:50140: read: connection reset by peer
‼ data plane and cli versions match
E0223 11:27:35.322654  156886 portforward.go:385] error copying from local connection to remote stream: read tcp4 127.0.0.1:39359->127.0.0.1:50142: read: connection reset by peer
    linkerd/linkerd-tap-568fdf6ddc-brq2h running stable-2.9.3 but cli running stable-2.9.2
    see https://linkerd.io/checks/#l5d-data-plane-cli-version for hints

linkerd-prometheus
------------------
√ prometheus add-on service account exists
√ prometheus add-on config map exists
√ prometheus pod is running

linkerd-grafana
---------------
√ grafana add-on service account exists
√ grafana add-on config map exists
√ grafana pod is running

Status check results are √
```

### Environment

- Kubernetes Version: v1.18.15-gke.1100
- Cluster Environment: GKE containerd://1.4.1
- Host OS: Container-Optimized OS from Google
- Linkerd version: stable-2.9.3

### Additional context
The invocation for the `proxy-init` library to `ConfigureFirewall` does not have an error handler to check the output as seen [here](https://github.com/linkerd/linkerd2/blob/f38ce611f03af8891e1e0d8604cdadabc388dce1/cni-plugin/main.go#L240). So if the firewall configuration process had errors, I'm not sure the main routine would catch it unless the program crashed entirely.

### PR DESCRIPTION
**Pull request by frankgu968**
This change adds error propagation for the CNI's ADD command so that any failures in the `ConfigureFirewall` function to configure the Pod's iptables can be bubbled up to be logged and handled.

Fixes #5809 

Signed-off-by: Frank Gu <frank@voiceflow.com>

### CODE REVIEW

File: cni-plugin/main.go [hunk 1]
```diff
@@ -237,7 +237,11 @@ func cmdAdd(args *skel.CmdArgs) error {
 				logEntry.Errorf("linkerd-cni: could not create a Firewall Configuration from the options: %v", options)
 				return err
 			}
-			iptables.ConfigureFirewall(*firewallConfiguration)
+
+			if err := iptables.ConfigureFirewall(*firewallConfiguration); err != nil {
```
Comment (cpretzer) [inline cni-plugin/main.go:241]: @frankgu968 thanks for the fast turnaround on this PR

The only thing I'd change is to move the `if err !=nil` check to it's own line, to be consistent with the rest of the checks, but that's trivial

Your contribution is much appreciated

**Review by cpretzer — APPROVED** [review]

File: cni-plugin/main.go (hunk 1 above)
Comment (frankgu968) [reply cni-plugin/main.go]: No problem! I saw the inline delaration pattern from earlier lines as well. Made the changes according to your suggestion @cpretzer

Comment (cpretzer) [reply cni-plugin/main.go]: thanks @frankgu968

**Review by kleimkuhler — APPROVED** [review]
Thanks @frankgu968!

**Pull request merged**